staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -160,12 +202,35 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 	}
 
 	data := originalData
-	if s.yaml {
-		altered, err := yaml.YAMLToJSON(data)
+	var yamlToJSONError error
+	if s.strict {
+		// In strict mode always pass the data trough the YAMLToJSONStrict converter.
+		// This is done to catch duplicate fields for both JSON and YAML. For JSON data,
+		// the output would equal the input, unless there is a parsing error such as duplicate fields.
+		altered, err := yaml.YAMLToJSONStrict(data)
 		if err != nil {
-			return nil, nil, err
+			// Store the error for later use, once a GVK for the input is found.
+			yamlToJSONError = err
+			if s.yaml {
+				// For YAML input, still process with the non-strict converter so that meta.Interpret()
+				// receives input converted to JSON.
+				altered, err := yaml.YAMLToJSON(data)
+				if err != nil {
+					return nil, nil, err
+				}
+				data = altered
+			}
+		} else {
+			data = altered
+		}
+	} else {
+		if s.yaml {
```

Comment:
I think this could be (maybe better) written as
```go
        var yamlToJSONError error
	if s.strict {
		// In strict mode always pass the data trough the YAMLToJSONStrict converter.
		// This is done to catch duplicate fields for both JSON and YAML. For JSON data,
		// the output would equal the input, unless there is a parsing error such as duplicate fields.
		altered, err := yaml.YAMLToJSONStrict(data)
		if err == nil {
                        // Update data with the sanitized, strict JSON. We update the data variable here, because the original input might have been YAML. For good JSON this is a no-op.
                        data = altered
		} else {
			// Store the error for later use, once a GVK for the input is detected so we can throw a good, structured error.
			yamlToJSONError = err
		}
	}
	// If this is the YAML decoder, and the codec is either non-strict, or have failed to decode in strict mode, convert the provided YAML to JSON in non-strict mode.
	if s.yaml && (!s.strict || yamlToJSONError != nil {
		altered, err := yaml.YAMLToJSON(data)
		if err != nil {
			return nil, nil, err
		}
		data = altered
	}
```

I think this might be easier to read, but IDK. I don't feel strongly